OpenRA.Game/Network/OrderManager.cs

```diff
@@ -126,9 +129,16 @@ void SendImmediateOrders()
 			localImmediateOrders.Clear();
 		}
 
-		public void ReceiveDisconnect(int clientIndex)
+		public void ReceiveDisconnect(int clientId, int frame)
 		{
-			pendingOrders.Remove(clientIndex);
+			// Queue a null order packet as a marker to run world/synced disconnect handlers deterministically
+			if (GameStarted)
+				ReceiveOrders(clientId, (frame, null));
```

Comment:
Without thinking all the way, could we have a hashset (or something similar) that we stored disconnected clients, and then we looked at that hashset in the ProcessOrders method? That would make this more explicit.

Reply:
The important change is that we track the frame as well as the client, and defer the removal until we process that frame.